File: python/src/roar_sdk/streaming.py
# -*- coding: utf-8 -*-
"""ROAR Protocol SDK — In-process event bus for Layer 5 streaming.

Zero external dependencies. Each subscriber gets a bounded asyncio queue.
Backpressure: when a queue is full, the oldest event is dropped.
"""

from __future__ import annotations

import asyncio
import logging
import uuid
from collections import deque
from dataclasses import dataclass, field
from typing import AsyncIterator, Deque, Dict, List, Optional

from .types import StreamEvent

logger = logging.getLogger(__name__)


@dataclass
class StreamFilter:
    """Filter for selecting which events a subscriber receives.

    All non-empty filters are AND-combined.
    """

    event_types: List[str] = field(default_factory=list)
    source_dids: List[str] = field(default_factory=list)
    session_ids: List[str] = field(default_factory=list)

    def matches(self, event: StreamEvent) -> bool:
        if self.event_types and event.type not in self.event_types:
            return False
        if self.source_dids and event.source not in self.source_dids:
            return False
        if self.session_ids and event.session_id not in self.session_ids:
            return False
        return True


class Subscription:
    """An active event subscription. Async-iterable."""

    def __init__(
        self,
        sub_id: str,
        filter_spec: StreamFilter,
        queue: asyncio.Queue,
        bus: "EventBus",
    ) -> None:
        self.id = sub_id
        self.filter = filter_spec
        self._queue = queue
        self._bus = bus
        self._closed = False
        self.events_received: int = 0
        self.events_dropped: int = 0

    @property
    def closed(self) -> bool:
        return self._closed

    async def __aiter__(self) -> AsyncIterator[StreamEvent]:
        while not self._closed:
            try:
                event = await asyncio.wait_for(self._queue.get(), timeout=1.0)
                self.events_received += 1
                yield event
            except asyncio.TimeoutError:
                continue

    async def get(self, timeout: float = 5.0) -> Optional[StreamEvent]:
        """Get the next event, or None on timeout."""
        if self._closed:
            return None
        try:
            event = await asyncio.wait_for(self._queue.get(), timeout=timeout)
            self.events_received += 1
            return event
        except asyncio.TimeoutError:
            return None

    def close(self) -> None:
        if not self._closed:
            self._closed = True
            self._bus._unsubscribe(self.id)


class EventBus:
    """In-process pub/sub event bus for ROAR streaming.

    Usage::

        bus = EventBus()
        sub = bus.subscribe(StreamFilter(event_types=["tool_call"]))
        await bus.publish(StreamEvent(type="tool_call", data={...}))
        event = await sub.get(timeout=2.0)
    """

    def __init__(self, max_buffer: int = 1000, replay_size: int = 100) -> None:
        self._max_buffer = max_buffer
        self._replay_buffer: Deque[StreamEvent] = deque(maxlen=replay_size)
        self._subscriptions: Dict[str, Subscription] = {}
        self._event_count: int = 0

    @property
    def subscriber_count(self) -> int:
        return len(self._subscriptions)

    @property
    def event_count(self) -> int:
        return self._event_count

    def subscribe(
        self,
        filter_spec: Optional[StreamFilter] = None,
        buffer_size: Optional[int] = None,
        replay: bool = False,
    ) -> Subscription:
        """Create a new subscription.

        Args:
            filter_spec: Event filter (None = receive all events).
            buffer_size: Per-subscriber buffer size.
            replay: Pre-fill with recent events from the replay buffer.
        """
        sub_id = f"sub-{uuid.uuid4().hex[:12]}"
        queue: asyncio.Queue = asyncio.Queue(maxsize=buffer_size or self._max_buffer)
        effective_filter = filter_spec or StreamFilter()

        if replay:
            for event in self._replay_buffer:
                if effective_filter.matches(event):
                    try:
                        queue.put_nowait(event)
                    except asyncio.QueueFull:
                        break

        sub = Subscription(sub_id=sub_id, filter_spec=effective_filter, queue=queue, bus=self)
        self._subscriptions[sub_id] = sub
        return sub

    async def publish(self, event: StreamEvent) -> int:
        """Publish an event to all matching subscribers. Returns delivery count."""
        self._event_count += 1
        self._replay_buffer.append(event)
        delivered = 0

        for sub in list(self._subscriptions.values()):
            if sub.closed or not sub.filter.matches(event):
                continue
            try:
                sub._queue.put_nowait(event)
                delivered += 1
            except asyncio.QueueFull:
                try:
                    sub._queue.get_nowait()
                    sub.events_dropped += 1
                except asyncio.QueueEmpty:
                    pass
                try:
                    sub._queue.put_nowait(event)
                    delivered += 1
                except asyncio.QueueFull:
                    sub.events_dropped += 1
                    logger.warning("Subscription %s dropping events (buffer full)", sub.id)

        return delivered

    def _unsubscribe(self, sub_id: str) -> None:
        self._subscriptions.pop(sub_id, None)

    def close_all(self) -> None:
        for sub in list(self._subscriptions.values()):
            sub._closed = True
        self._subscriptions.clear()

File: ts/src/streaming.ts
/**
 * ROAR Protocol — in-process EventBus for Layer 5 streaming.
 *
 * Mirrors Python's roar_sdk/streaming.py.
 * Zero external dependencies — uses Node.js built-in EventEmitter as a signal only.
 */

import { EventEmitter } from "events";
import { StreamEvent, StreamEventType } from "./types.js";

// ---------------------------------------------------------------------------
// StreamFilter
// ---------------------------------------------------------------------------

export interface StreamFilterSpec {
  event_types?: StreamEventType[];
  source_dids?: string[];
  session_ids?: string[];
}

export class StreamFilter {
  readonly event_types: StreamEventType[];
  readonly source_dids: string[];
  readonly session_ids: string[];

  constructor(spec: StreamFilterSpec = {}) {
    this.event_types = spec.event_types ?? [];
    this.source_dids = spec.source_dids ?? [];
    this.session_ids = spec.session_ids ?? [];
  }

  matches(event: StreamEvent): boolean {
    if (this.event_types.length > 0 && !this.event_types.includes(event.type))
      return false;
    if (this.source_dids.length > 0 && !this.source_dids.includes(event.source))
      return false;
    if (
      this.session_ids.length > 0 &&
      !this.session_ids.includes(event.session_id)
    )
      return false;
    return true;
  }
}

// ---------------------------------------------------------------------------
// Subscription
// ---------------------------------------------------------------------------

export class Subscription implements AsyncIterable<StreamEvent> {
  readonly id: string;
  readonly filter: StreamFilter;

  private _buffer: StreamEvent[] = [];
  private _closed = false;
  private _resolve: ((e: StreamEvent) => void) | null = null;
  private _bus: EventBus;

  eventsReceived = 0;
  eventsDropped = 0;

  // AIMD capacity: floats between 1 and maxBuffer
  private _capacity: number;

  constructor(
    id: string,
    filter: StreamFilter,
    bus: EventBus,
    private readonly maxBuffer: number,
  ) {
    this.id = id;
    this.filter = filter;
    this._bus = bus;
    this._capacity = maxBuffer;
  }

  /** AIMD effective limit (floor of floating capacity). */
  get effectiveLimit(): number {
    return Math.floor(this._capacity);
  }

  /** @internal — called by EventBus.publish */
  _deliver(event: StreamEvent): boolean {
    if (this._closed) return false;
    if (this._resolve) {
      // Waiting consumer — deliver directly; additive increase
      const resolve = this._resolve;
      this._resolve = null;
      this._capacity = Math.min(this.maxBuffer, this._capacity + 1);
      this.eventsReceived++;
      resolve(event);
      return true;
    }
    if (this._buffer.length >= this.effectiveLimit) {
      // Backpressure: multiplicative decrease, drop oldest
      this._capacity = Math.max(1, this._capacity * 0.5);
      this._buffer.shift();
      this.eventsDropped++;
    } else {
      // Buffered without drop: additive increase
      this._capacity = Math.min(this.maxBuffer, this._capacity + 1);
    }
    this._buffer.push(event);
    return true;
  }

  /** Get the next event, or undefined after timeout (ms). */
  async get(timeoutMs = 5000): Promise<StreamEvent | undefined> {
    if (this._closed) return undefined;
    if (this._buffer.length > 0) {
      this.eventsReceived++;
      return this._buffer.shift()!;
    }
    return new Promise<StreamEvent | undefined>((resolve) => {
      this._resolve = (e) => resolve(e);
      setTimeout(() => {
        if (this._resolve === resolve) {
          this._resolve = null;
          resolve(undefined);
        }
      }, timeoutMs);
    });
  }

  close(): void {
    if (!this._closed) {
      this._closed = true;
      this._resolve?.(undefined as unknown as StreamEvent);
      this._resolve = null;
      this._bus._unsubscribe(this.id);
    }
  }

  get closed(): boolean {
    return this._closed;
  }

  async *[Symbol.asyncIterator](): AsyncGenerator<StreamEvent> {
    while (!this._closed) {
      const event = await this.get(1000);
      if (event !== undefined) yield event;
    }
  }
}

// ---------------------------------------------------------------------------
// EventBus
// ---------------------------------------------------------------------------

export class EventBus {
  private _subscriptions = new Map<string, Subscription>();
  private _replayBuffer: StreamEvent[] = [];
  private _eventCount = 0;
  private _counter = 0;

  constructor(
    private readonly maxBuffer = 1000,
    private readonly replaySize = 100,
  ) {}

  get subscriberCount(): number {
    return this._subscriptions.size;
  }

  get eventCount(): number {
    return this._eventCount;
  }

  subscribe(
    filterSpec?: StreamFilterSpec,
    bufferSize?: number,
    replay = false,
  ): Subscription {
    const id = `sub-${(++this._counter).toString(16).padStart(12, "0")}`;
    const filter = new StreamFilter(filterSpec);
    const sub = new Subscription(id, filter, this, bufferSize ?? this.maxBuffer);

    if (replay) {
      for (const event of this._replayBuffer) {
        if (filter.matches(event)) sub._deliver(event);
      }
    }

    this._subscriptions.set(id, sub);
    return sub;
  }

  /** Publish an event to all matching subscribers. Returns delivery count. */
  publish(event: StreamEvent): number {
    this._eventCount++;
    this._replayBuffer.push(event);
    if (this._replayBuffer.length > this.replaySize) {
      this._replayBuffer.shift();
    }

    let delivered = 0;
    for (const sub of this._subscriptions.values()) {
      if (!sub.closed && sub.filter.matches(event)) {
        if (sub._deliver(event)) delivered++;
      }
    }
    return delivered;
  }

  _unsubscribe(id: string): void {
    this._subscriptions.delete(id);
  }

  closeAll(): void {
    for (const sub of this._subscriptions.values()) {
      sub.close();
    }
    this._subscriptions.clear();
  }
}
